src/edu/washington/escience/myria/operator/Sample.java

```diff
@@ -0,0 +1,241 @@
+        samples = generateIndicesWR(streamSize, sampleSize);
+      } else if (sampleType.equals("WoR")) {
+        samples = generateIndicesWoR(streamSize, sampleSize);
+      } else {
+        throw new DbException("Invalid sampleType: " + sampleType);
+      }
+
+      computedSamplingInfo = true;
+    }
+
+    // Return a ready tuple batch if possible.
+    TupleBatch nexttb = ans.popAny();
+    if (nexttb != null) {
+      return nexttb;
+    }
+    // Check if there's nothing left to sample.
+    if (curSampIdx >= samples.length) {
+      getRight().close();
+      setEOS();
+      return null;
+    }
+    Operator right = getRight();
+    for (TupleBatch tb = right.nextReady(); tb != null; tb = right.nextReady()) {
+      if (curSampIdx >= samples.length) { // done sampling
+        break;
+      }
+      if (samples[curSampIdx] > tupleNum + tb.numTuples()) {
+        // nextIndex is not in this batch. Continue with next batch.
+        tupleNum += tb.numTuples();
+        continue;
+      }
+      while (curSampIdx < samples.length
+          && samples[curSampIdx] < tupleNum + tb.numTuples()) {
+        ans.put(tb, samples[curSampIdx] - tupleNum);
+        curSampIdx++;
+      }
+      tupleNum += tb.numTuples();
+      if (ans.hasFilledTB()) {
+        return ans.popFilled();
+      }
+    }
+    return ans.popAny();
+  }
+
+  /** Helper function to extract sampling information from a TupleBatch. */
+  private void extractSamplingInfo(TupleBatch tb) throws Exception {
+    Preconditions.checkArgument(tb != null);
+
+    int workerID;
+    Type col0Type = tb.getSchema().getColumnType(0);
+    if (col0Type == Type.INT_TYPE) {
+      workerID = tb.getInt(0, 0);
+    } else if (col0Type == Type.LONG_TYPE) {
+      workerID = (int) tb.getLong(0, 0);
+    } else {
+      throw new DbException("WorkerID column must be of type INT or LONG");
+    }
+    Preconditions.checkState(workerID == getNodeID(),
+            "Invalid WorkerID for this worker. Expected %s, but received %s",
+            getNodeID(), workerID);
+
+    Type col1Type = tb.getSchema().getColumnType(1);
+    if (col1Type == Type.INT_TYPE) {
+      streamSize = tb.getInt(1, 0);
+    } else if (col1Type == Type.LONG_TYPE) {
+      streamSize = (int) tb.getLong(1, 0);
+    } else {
+      throw new DbException("StreamSize column must be of type INT or LONG");
+    }
+    Preconditions.checkState(streamSize >= 0, "streamSize cannot be negative");
+
+    Type col2Type = tb.getSchema().getColumnType(2);
+    if (col2Type == Type.INT_TYPE) {
+      sampleSize = tb.getInt(2, 0);
+    } else if (col2Type == Type.LONG_TYPE) {
+      sampleSize = (int) tb.getLong(2, 0);
+    } else {
+      throw new DbException("SampleSize column must be of type INT or LONG");
+    }
+    Preconditions.checkState(sampleSize >= 0, "sampleSize cannot be negative");
+
+    Type col3Type = tb.getSchema().getColumnType(3);
+    if (col3Type == Type.STRING_TYPE) {
+      sampleType = tb.getString(3, 0);
```

Comment:
If we change `SampleType` to an enum then we should catch an exception from parsing an unknown string value and throw the appropriate `DbException`.